### PR TITLE
AG-7879 - Modularised navigator + pre-work for enterprise modules

### PR DESCRIPTION
**Pull request by alantreadway**
https://ag-grid.atlassian.net/browse/AG-7879

Add interface contracts for enterprise modules, and improve the coupling of the navigator feature through abstraction of the layout & zoom management.

### CODE REVIEW

File: charts-packages/ag-charts-community/src/chart/cartesianChart.ts [hunk 1]
```diff
@@ -45,28 +40,21 @@ export class CartesianChart extends Chart {
             shrinkRect.shrink(legendPadding, legend.position);
         }
 
-        if (navigator.enabled) {
-            const navigatorTotalHeight = navigator.height + navigator.margin;
-            shrinkRect.shrink(navigatorTotalHeight, 'bottom');
-            navigator.y = shrinkRect.y + shrinkRect.height + navigator.margin;
-        }
+        ({ shrinkRect } = this.layoutService.dispatchPerformLayout('before-series', { shrinkRect }));
```
Comment (alexanderby) [inline charts-packages/ag-charts-community/src/chart/cartesianChart.ts:43]: image

File: charts-packages/ag-charts-community/src/chart/chart.ts [hunk 2]
```diff
@@ -294,6 +344,12 @@ export abstract class Chart extends Observable implements AgChartInstance {
         this.tooltip.destroy();
         SizeMonitor.unobserve(this.element);
 
+        for (const [key, module] of Object.entries(this.modules)) {
+            module.instance.destroy();
+            delete this.modules[key];
+            (this as any)[key];
```
Comment (alexanderby) [inline charts-packages/ag-charts-community/src/chart/chart.ts:350]: What does this line do?

Comment (alexanderby) [inline charts-packages/ag-charts-community/src/chart/chart.ts:350]: What does this line do?

**Review by alexanderby — APPROVED** [review]

File: charts-packages/ag-charts-community/src/chart/chart.ts (hunk 2 above)
Comment (alantreadway) [reply charts-packages/ag-charts-community/src/chart/chart.ts:350]: It (hackily) sets the property `navigator` on the `Chart` instance - the factory can then apply options as before.

We may want to revisit this in the future.

Comment (alantreadway) [reply charts-packages/ag-charts-community/src/chart/chart.ts:350]: Fixed in https://github.com/ag-grid/ag-grid/commit/69bb8c07d1d2d6966700c6982ee1f7ff2b43d948

**Pull request merged**